src/p11_wrap.cpp

```diff
@@ -0,0 +1,276 @@
+ * limitations under the License.
+ * #L%
+ */
+
+/*
+ * This file is the only place where we should see any
+ * methods from LibP11. Any other code should use
+ * the more high-level methods declared and define in
+ * p11_wrap.cpp and p11_wrap.h.
+ *
+ */
+
+#include <boost/algorithm/hex.hpp>
+#include <cassert>
+#include <cstddef> /* this has to come before cppc (bug in boost) */
+#include <exception>
+
+#include <cppc/checkcall.hpp>
+#include "mococrw/openssl_wrap.h"
+#include "mococrw/p11_wrap.h"
+
+namespace mococrw
+{
+namespace p11
+{
+// Initialise error strings of LibP11 wrapper:
+const std::string P11Exception::nullCtxExceptionString = "Found NULL PKCS11 Context";
+const std::string P11Exception::nullSlotListExceptionString = "Found NULL Slot List";
+const std::string P11Exception::nullTokenExceptionString = "Found NULL Token";
+const std::string P11Exception::mismatchLoginExceptionString = "Login/Logout Mismatch Detected";
+
+/**
+ * This struct is used by CWrap to
+ * determine how to react to an error
+ * when performing a LibP11 call.
+ */
+struct P11ExceptionErrorPolicy
+{
+    template <class Rv>
+    static void handleError(const Rv &);
+};
+
+/**
+ * Throw a P11 exception upon error.
+ *
+ * This method fetches the error string and throws an exception
+ * with the corresponding message.
+ */
+template <class Rv>
+void P11ExceptionErrorPolicy::handleError(const Rv & /* unused */)
+{
+    throw P11Exception();
+}
+
+// For checking return values:
+using P11CallPtr =
+        ::cppc::CallCheckContext<::cppc::IsNotNullptrReturnCheckPolicy, P11ExceptionErrorPolicy>;
+using P11IsNonNegative =
+        ::cppc::CallCheckContext<::cppc::IsNotNegativeReturnCheckPolicy, P11ExceptionErrorPolicy>;
+
+/***********************************************************************************
+ * Implementation of P11_SlotInfo
+ */
+
+P11_SlotInfo::P11_SlotInfo(P11_PKCS11_CTX_Ptr ctx) : _ctx(ctx), _slotList(nullptr), _numSlots(0)
+{
+    PKCS11_SLOT *slots;
+
+    // Enumerate slots to obtain slot information.
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_enumerate_slots, _ctx.get(), &slots, &_numSlots);
+
+    // If the returned slot list is NULL, we cannot continue!
+    if (slots == nullptr) {
+        throw P11Exception(P11Exception::nullSlotListExceptionString);
+    }
+
+    SlotListDeleter deleter(_ctx, _numSlots);
+    _slotList = P11_PKCS11_SLOT_LIST_PTR(slots, deleter);
+}
+
+void P11_SlotInfo::SlotListDeleter::operator()(PKCS11_SLOT *slotList)
+{
+    if (slotList == nullptr) {
+        throw P11Exception(P11Exception::nullSlotListExceptionString);
+    }
+
+    lib::LibP11::P11_PKCS11_release_all_slots(_ctx.get(), slotList, _numSlots);
+}
+
+P11_PKCS11_SLOT_PTR P11_SlotInfo::findSlot()
+{
+    PKCS11_SLOT *raw_slot = P11CallPtr::callChecked(
+            lib::LibP11::P11_PKCS11_find_token, _ctx.get(), _slotList.get(), _numSlots);
+
+    // Leverage aliasing.
+    return P11_PKCS11_SLOT_PTR(_slotList, raw_slot);
```

Comment:
This is the shared pointer's aliasing constructor. Basically since `raw_slot` is a pointer to memory owned by `_slotList`, the alias constructor is used to add the new shared_pointer, associated to the `raw_slot` pointer, in the share group containing `_slotList`.

See https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr